test/e2e/network/topology.go

```diff
@@ -0,0 +1,205 @@
+			}
+		}, int32(port))
+		framework.ExpectNoError(err)
+		svcIP := svc.Spec.ClusterIP
+
+		ginkgo.By("creating backend pods for the service on each node" + serviceName)
+		// hints are only generated if there is a reasonable distribution of endpoints
+		// TODO(aojea): revisit, this number was obtained with KIND in a cluster with 3 nodes
+		err = jig.CreateServicePods(3 * len(nodes.Items))
+		framework.ExpectNoError(err)
+
+		// check that topology hints are created
+		opts := metav1.ListOptions{
+			LabelSelector: "kubernetes.io/service-name=" + serviceName,
+		}
+
+		// map that contains the zone associated to each pod
+		podsZones := map[string]string{}
+		err = wait.PollImmediate(3*time.Second, 1*time.Minute, func() (bool, error) {
+			es, err := cs.DiscoveryV1().EndpointSlices(namespace).List(context.TODO(), opts)
+			if err != nil {
+				framework.Logf("Failed to get list EndpointSlice objects: %v", err)
+				// Retry the error
+				return false, nil
+			}
+			for _, endpointSlice := range es.Items {
+				for _, ep := range endpointSlice.Endpoints {
+					if ep.Hints == nil {
+						framework.Logf("Failed to get Topology hints on endpoint")
+						return false, nil
+					}
+					podsZones[ep.TargetRef.Name] = *ep.Zone
+					framework.Logf("pod %s on zone %s with hints %v", ep.TargetRef.Name, *ep.Zone, ep.Hints)
+				}
+			}
+			return true, nil
+
+		})
+		framework.ExpectNoError(err)
+
+		// create a pod in a specific zone to check the traffic distribution
+		execPod := e2epod.CreateExecPodOrFail(cs, namespace, "execpod", func(pod *v1.Pod) {
+			pod.Spec.NodeName = nodeName
+		})
+		defer func() {
+			framework.Logf("Cleaning up the exec pod")
+			err := cs.CoreV1().Pods(namespace).Delete(context.TODO(), execPod.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete pod: %s in namespace: %s", execPod.Name, namespace)
+		}()
+		err = jig.CheckServiceReachability(svc, execPod)
+		framework.ExpectNoError(err)
+
+		// poll the service 100 times and get a per zone traffic distribution
+		zones := map[string]int{}
+		maxConnections := 100
+		nc := fmt.Sprintf(`echo hostName | nc -v -w 5 %s %d`, svcIP, port)
+		cmd := fmt.Sprintf("for i in $(seq 0 %d); do echo; %s ; done", maxConnections, nc)
+		stdout, err := framework.RunHostCmd(execPod.Namespace, execPod.Name, cmd)
+		framework.ExpectNoError(err)
+
+		hostnames := strings.Split(stdout, "\n")
+		failed := 0
+		// failed request return an empty string that we can not associate to a zone
+		for _, h := range hostnames {
+			z, ok := podsZones[h]
+			if ok {
+				zones[z]++
+			} else {
+				failed++
+			}
+		}
+		framework.Logf("Connections from Zone %s distributed with topology hints enabled %v Failed: %d", nodeZone, zones, failed)
+		// Fail if the traffic in the zone is not higher than the 65%
+		// don't consider the failed connections for the stats
+		// TODO(aojea): based on local testing with KIND revisit percentages.
```

Comment:
This feels like it's going to be very flaky unless I'm missing test details. Are we doing anything above to ensure that pods are distributed evenly across nodes/zones? If not, it seems like we'd constantly run into scenarios where all pods were on the same node and this test failed. It seems like a safer approach would be to have 3 equivalent nodes in each zone + deploy pods evenly across those nodes and then ensure that 100% of requests from a zone stayed in that zone.

Reply:
let's talk in https://github.com/kubernetes/kubernetes/pull/100807